### PR TITLE
[SPARK-23961][SPARK-27548][PYTHON] Fix error when toLocalIterator goes out of scope and properly raise errors from worker

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

This fixes an error when a PySpark local iterator, for both RDD and DataFrames, goes out of scope and the connection is closed before fully consuming the iterator. The error occurs on the JVM in the serving thread, when Python closes the local socket while the JVM is writing to it. This usually happens when there is enough data to fill the socket read buffer, causing the write call to block.

Additionally, this fixes a problem when an error occurs in the Python worker and the collect job is cancelled with an exception. Previously, the Python driver was never notified of the error so the user could get a partial result (iteration until the error) and the application will continue. With this change, an error in the worker is sent to the Python iterator and is then raised.

The change here introduces a protocol for PySpark local iterators that work as follows:

1) The local socket connection is made when the iterator is created
2) When iterating, Python first sends a request for partition data as a non-zero integer
3) While the JVM local iterator over partitions has next, it triggers a job to collect the next partition
4) The JVM sends a nonzero response to indicate it has the next partition to send
5) The next partition is sent to Python and read by the PySpark deserializer
6) After sending the entire partition, an `END_OF_DATA_SECTION` is sent to Python which stops the deserializer and allows to make another request
7) When the JVM gets a request from Python but has already consumed it's local iterator, it will send a zero response to Python and both will close the socket cleanly
8) If an error occurs in the worker, a negative response is sent to Python followed by the error message. Python will then raise a RuntimeError with the message, stopping iteration.
9) When the PySpark local iterator is garbage-collected, it will read any remaining data from the current partition (this is data that has already been collected) and send a request of zero to tell the JVM to stop collection jobs and close the connection.

Steps 1, 3, 5, 6 are the same as before. Step 8 was completely missing before because errors in the worker were never communicated back to Python. The other steps add synchronization to allow for a clean closing of the socket, with a small trade-off in performance for each partition. This is mainly because the JVM does not start collecting partition data until it receives a request to do so, where before it would eagerly write all data until the socket receive buffer is full.

## How was this patch tested?

Added new unit tests for DataFrame and RDD `toLocalIterator` and tested not fully consuming the iterator. Manual tests with Python 2.7  and 3.6.

### CODE REVIEW

**Comment by BryanCutler** [comment]
### Timings for DataFrame.toLocalIterator and RDD.toLocalIterator

These tests are to illustrate the slowdown caused by this change, comparing current master with this change. Wall clock time is measure to fully consume the local iterator and average of 5 runs are shown:

 _ | DataFrame | RDD
--- | --- | ---
*master* | 10.26016583 | 4.354181528
*this PR* | 12.14033799 | 3.823320436

#### Test Script

```python
import time
from pyspark.sql import SparkSession

spark = SparkSession\
        .builder\
        .appName("toLocalIterator_timing")\
        .getOrCreate()

num = 1 << 22
numParts = 32

def run(df):
  print("Starting iterator:")
  start = time.time()

  count = 0
  for row in df.toLocalIterator():
    count += 1

  if count != num:
    raise RuntimeError("Expected {} but got {}".format(num, count))

  elapsed = time.time() - start
  print("completed in {}".format(elapsed))

run(spark.range(num, numPartitions=numParts))
run(spark.sparkContext.range(num, numSlices=numParts))

spark.stop()


```

**Comment by BryanCutler** [comment]
I just want to highlight that the error that this fixes only kill the serving thread and Spark can continue normal operation. Although the error is pretty ugly and would lead users to think that something went terribly wrong. Since it's pretty common to not fully consume an iterator, e.g. taking a slice, I believe it is worth making this change.

It is also possible that this change would be very beneficial because if the iterator is not fully consumed, it could save the triggering of unneeded jobs where the behavior before eagerly queued jobs for all partitions. In this sense, the change here more closely follows the Scala behavior.

I'm also not entirely sure why I'm seeing a speedup for the RDD toLocalIterator. When using 8 partitions instead of 32, I noticed a slowdown. I will try to run some more tests.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (BryanCutler) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Note: this is the same function as `collectPartition(p: Int)` in Scala, except here we do not want to flatten the collected arrays

File: python/pyspark/rdd.py
Comment (BryanCutler) [inline python/pyspark/rdd.py]: TODO: need to address this

File: python/pyspark/sql/tests/test_dataframe.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_dataframe.py]: Note: this would not have crashed before, only generated the error and I don't think it's possible to check if this error happened

File: python/pyspark/tests/test_rdd.py
Comment (BryanCutler) [inline python/pyspark/tests/test_rdd.py]: This sleep is unnecessary. `rdd.toLocalIterator` makes the socket connection and iterating starts reading from the pyspark serializer.

File: python/pyspark/tests/test_rdd.py
Comment (BryanCutler) [inline python/pyspark/tests/test_rdd.py]: ditto

**Comment by BryanCutler** [comment]
Ping @HyukjinKwon @holdenk @viirya for thoughts on this, thanks!

**Comment by BryanCutler** [comment]
gentle ping @HyukjinKwon @ueshin for thoughts on this fix

File: python/pyspark/rdd.py
Comment (BryanCutler) [reply python/pyspark/rdd.py]: ah no, that was leftover from a previous revision.  I'll remove that and clean up.

File: python/pyspark/rdd.py
Comment (BryanCutler) [reply python/pyspark/rdd.py]: yeah, good idea thanks!

**Comment by BryanCutler** [comment]
Thanks @ueshin , I updated the code.  I also made a change for the JVM to send a response to indicate if there are more partitions to read instead of relying on closing the socket and catching an error, which could have masked real errors. This change did not have any significant difference to the timings above. Please take another look when you can.

**Comment by holdenk** [comment]
So before I jump into the details of this PR, from the design side I've got some questions I'd like to try and understand better:

1) This description mentions that it may save "unneeded work" but not eagerly queueing jobs that aren't needed, but in the situation where the jobs are needed it seems like we're adding unnecessary synchronization that could slow this down.
- Have you tried the benchmarks with a slow per-partition computation to see if there is more impact there?
- Is this slow down "worth it", 20% on DataFrame is nothing to walk away from if we can avoid it.
- If we're worried about the driver program falling over from buffering  too much data in the JVM can we maybe queue the next partition while we are serving the current one to decrease the overhead?

2) Are we solving this problem on the "right" side? Is it common for the the localiterator to go out of scope before being consumed?
- If it is is a relatively infrequent occurrence, would it possible make sense to just make the Java code accept that the Python socket may go away? I think that might keep the "happy" (or more frequent) path at reasonable performance. On the other hand if folks are using toLocalIterator to partially consume the data and this is the expected path, then using exception handling might not make sense.

I'm going to dig into the code, but I'd love to get a better understanding of the design as well.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: For performance, as mentioned in my questions, would it make sense to use something like a iterator with look ahead of say 1 partition (or X% of partitions) so we decrease the blocking time.

**Review by holdenk — COMMENTED** [review]
Thanks for working on this issue, I'd love to get a better understanding of the design trade-offs but really excited to see this get fixed and also exited to see less `sleep(5)` in the code base :) :p

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (BryanCutler) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: No, the remaining jobs are not triggered. The python iterator finishes consuming the data from the current job, then sends a command for Scala iterator to stop.

**Comment by BryanCutler** [comment]
> Are we solving this problem on the "right" side? Is it common for the the localiterator to go out of scope before being consumed?
If it is is a relatively infrequent occurrence, would it possible make sense to just make the Java code accept that the Python socket may go away?

Thanks @holdenk , I totally agree.  I started off with a fix that would catch the error on the Java side and just allow the connection to be broken without showing the error.  Let me summarize the trade-offs that I noticed:

**Catch/Ignore error in Java:**
- No change in the protocol to Python, keeps things the same
- This is the fastest way to iterate through all elements
- Ignoring the error _might_ hide an actual error, but this is over a local socket so once the connection is made I'm not sure how likely an error is
- While Python is still consuming data from one partition, other jobs will be started to collect the next partitions until write to the socket is blocked again.  This is nice if all the data is used, but bad if it's not.

**Synchronized Protocol (this PR):**
- Allows for the iterator to go out of scope and close the connection cleanly, without needing to catch/ignore errors.
- Jobs are only triggered when the next element is needed. So only after the last element of  a partition is consumed, the next partition is collected. This matches the behavior of Scala toLocalIterator.
- Performance is a little slower because of synchronization and no longer eagerly collecting the next partition
- Makes the communication between Python more complicated

I went with the second option here because I definitely didn't want to mask potential real errors, and only triggering jobs for data that is requested, matches Scala and is good for some use cases. Also while we don't want performance to suck, I think people would use toLocalIterator to work with constrained memory (1 partition at a time) and do a regular collect if they want good performance.

Btw, from the JIRA for this, the user was calling `itertools.islice(..)` to take a slice of the the local iterator, which wasn't fully consuming it. I would agree it is probably more common to consume the entire iterator though.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (BryanCutler) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Yeah, that would have better performance, but it does say in the doc that max memory usage will be the largest partition. Going over that might cause problems for some people, no?

File: python/pyspark/sql/tests/test_dataframe.py
Comment (BryanCutler) [reply python/pyspark/sql/tests/test_dataframe.py]: Yeah, that might work. I can give it a try..

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (viirya) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: How about previous behavior? The behavior before will trigger them? Looks like `toLocalIterator` won't trigger the job if we don't iterate the data on a partition.

Comment (BryanCutler) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: The previous behavior was that the Scala local iterator would advance as long as the `write` calls to the socket are not blocked. So this means when Python reads a batch (auto-batched elements) from the current partition, this will unblock the Scala call to write and could start the next job.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Once the local iterator at Python side is out of scope and so the iterator is not fully consumed, will it block the write call at Scala? Seems to me that it will and we shouldn't see unneeded jobs to be triggered after that, doesn't?

Comment (BryanCutler) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: The previous behavior is when the iterator goes out of scope, the socket eventually is closed. This creates the error on the Scala side and the writing thread is terminated, so no more jobs are triggered but the user sees this error.

**Comment by holdenk** [comment]
Thanks for the additional context @BryanCutler that really helps.

I think supporting memory constrained consumption of a large dataset is core to the goal of toLocalIterator so while there is a performance penality of this change, and we can work to minimize it, I think it's the right thing to do.

I agree with you solving it in the Python side seems like the right set of trade-offs. I think it _might_ make sense to kick off the job for the next partition (e.g. lookahead of 1), but we should totally do that in a follow up PR/JIRA as an optiimization. What do you think?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: That's a good point @BryanCutler, we could implement the lookahead as a seperate PR/JIRA and allow it to be turned-on/off. I'd suggest this PR is more about fixing the behaviour of toLocalIterator memory wise than the out-of-scope issue (although the out-of-scope issue is maybe more visible in the logs).

**Review by holdenk — COMMENTED** [review]
Thanks for the additional context, I agree with the design I think it is the correct set of trade-offs. I think it would be good to make a follow up JIRA for trying to improve the performance though if you agree.

**Comment by BryanCutler** [comment]
Thanks @holdenk ! I think having a lookahead for 1 partition makes sense as an option and that should bring the performance back up to where it was before. Sounds good to me to do this as a followup also. So do you think this is ok to merge? Let me rebase and test once more..

File: python/pyspark/sql/tests/test_dataframe.py
Comment (holdenk) [reply python/pyspark/sql/tests/test_dataframe.py]: Did anything come of this? It's optional so don't block on it.

File: python/pyspark/tests/test_rdd.py
Comment (holdenk) [reply python/pyspark/tests/test_rdd.py]: I like less sleeps in the code <3

**Comment by holdenk** [comment]
Provided the AppVeyor failure is unrelated I think it's ok to merge, it would be nice to have a more explicit test for this but capturing the JVM output isn't something we do elsewhere so I'm OK with putting that off (although I'd like to see a JIRA for it maybe?). Thanks for working on this -- if you do make the follow up issues do CC me on them :)

File: python/pyspark/sql/tests/test_dataframe.py
Comment (BryanCutler) [reply python/pyspark/sql/tests/test_dataframe.py]: Oh right, I forgot about this. Let me give it a shot now.

Comment (BryanCutler) [reply python/pyspark/sql/tests/test_dataframe.py]: It ends up being a little complicated, maybe better to try as a followup

**Comment by BryanCutler** [comment]
> it would be nice to have a more explicit test for this but capturing the JVM output isn't something we do elsewhere so I'm OK with putting that off (although I'd like to see a JIRA for it maybe?)

Thanks @holdenk ! I was close to getting a better test that captures JVM output using `launch_gateway` with `popen_kwargs` but there are a couple of tricky things doing it this way so I think it would be better to try as a followup where we could discuss more.

**Comment by BryanCutler** [comment]
test this please

**Comment by holdenk** [comment]
Sounds good @BryanCutler if your @ Spark Summit this week we can catch up in person about the difficult with the tests otherwise happy to chat whenever.

**Comment by holdenk** [comment]
I have another idea on how we can verify this behaviour change: parallelize a collection with 2 partitions and do map which throws on the last element and do a toLocalIterator and in the old version it _should_ fail but in the new version it should not fail. What do you think about that as a test?

**Comment by BryanCutler** [comment]
@holdenk yeah I will be at Spark Summit and it would be great to meetup! 

>do map which throws on the last element and do a toLocalIterator and in the old version it should fail 

Good idea! That will definitely work, I'll add that now

**Comment by BryanCutler** [comment]
Uggh, so the new test raises an error in the worker (running the old version) but it never goes through Py4J and the python driver never knows it happens. So the test is pretty much useless too. I made https://issues.apache.org/jira/browse/SPARK-27548 for this problem, I can work on that and then should be able to make a real test for this PR.

**Comment by BryanCutler** [comment]
It ended up being a lot easier to fix https://issues.apache.org/jira/browse/SPARK-27548 here, so I just added that here as well.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (BryanCutler) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: We want to catch any errors during the collection job, so I believe the `SparkException` should all that is needed here?

**Comment by viirya** [comment]
retest this please.

**Comment by BryanCutler** [comment]
I added the fix for https://issues.apache.org/jira/browse/SPARK-27548 here and updated the PR description. Basically when an error occurs on a worker, the collect job fails in the JVM and the error is sent to the Python iterable, which raises the error. If you could please have another look @holdenk @viirya @ueshin @HyukjinKwon and see if you agree with the approach. Thanks!

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: I think that's reasonable +1

**Review by holdenk — APPROVED** [review]
LGTM pending others comments of course, although if no one else has anything to say lets go with it :)

**Comment by BryanCutler** [comment]
Thanks @holdenk ! I will merge this after the current tests pass if there are no further comments.

**Comment by BryanCutler** [comment]
retest this please

**Comment by BryanCutler** [comment]
merged to master, thanks all for reviewing

**Comment by HyukjinKwon** [comment]
Sorry for my late response. I didn't check super closely but looks a-okay to me too.

**Comment by BryanCutler** [comment]
I made https://issues.apache.org/jira/browse/SPARK-27660 to explore prefetching data

**Pull request closed without merging**
